Azure snapshots

kensipe opened the pull request:
Currently uploads and downloads to azure and s3.  The REST endpt is there as well.

Compression currently isn't implemented.

CLI added!

test with CLI

```
dcos cassandra backup start \
   --backup_name=ken2 \
   --external_location=azure://cli2-backup \
   --azure_account=sipestore \
   --azure_key=itaSecret
```

Invocations to test

```
// changes for rest (azure)
curl -H "Content-Type: application/json" -H "Authorization: token=$(dcos config show core.dcos_acs_token)" -X PUT -d '{"backup_name":"kena","external_location":"azure://curl-backup","azure_account":"sipestore","azure_key":"itsASecret"}' http://ken-ojfao-elasticl-1ip1xlsweia3s-1034279044.us-west-2.elb.amazonaws.com/service/cassandra/v1/backup/start

// changes (s3)
curl -H "Content-Type: application/json" -H "Authorization: token=$(dcos config show core.dcos_acs_token)" -X PUT -d '{"backup_name":"ken3","external_location":"s3://azure-cass","s3_access_key":"AKIAJFSUBIZBVWERNOVQ","s3_secret_key":"stillASecret"}' http://ken-ojfao-elasticl-1ip1xlsweia3s-1034279044.us-west-2.elb.amazonaws.com/service/cassandra/v1/backup/start
```

test universe: https://github.com/mesosphere/universe-1.7/archive/cassandra-azure.zip

kensipe committed:
added the azure libraries and configured the storage factory

added azure storage logic for upload and refactored S3 to share common code

adding in download and some code cleanup

added logging to determine which driver was selected

code cleanup

the largest number of changes to a single project I have ever made to accommodate 2 properties.  2 properties.

need restore REST end pt

cli changes to accommodate azure backups

directory where downloads go doesn't exist prior to be writing files.  added better error handling on restore.

directories must be created first.

added additional documentation for restoring from azure

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

kensipe committed:
added compression capabilities to the azure storage driver which required the implementation of streams for page blobs

mohitsoni commented on cassandra-executor/src/main/java/com/mesosphere/dcos/cassandra/executor/backup/AzureStorageDriver.java:
```diff
@@ -0,0 +1,303 @@
+    CloudBlobContainer container = null;
+
+    if (StringUtils.isNotBlank(containerName)) {
+      final String storageConnectionString = "DefaultEndpointsProtocol=http"
```
Can this ever be `https` ? If yes, is there any reason to not make it https by default ?

mohitsoni commented on cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/resources/StartBackupRequest.java:
```diff
@@ -4,56 +4,77 @@
+
+  public boolean isValid() {
+    return name != null && externalLocation != null &&
+      (s3AccessKey != null && s3SecretKey != null) ||
```
I like the externalLocation check. Maybe we should add `externalLocation.startsWith("s3:")`

kensipe replied:
fixed plus added StartBackupRequestSpec to test combinations. (which resulted in a bug fix)

mohitsoni commented on cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/resources/StartRestoreRequest.java:
```diff
@@ -52,8 +58,25 @@ public void setS3SecretKey(String s3SecretKey) {
+
+    public boolean isValid() {
+      return name != null && externalLocation != null &&
+        (s3AccessKey != null && s3SecretKey != null) ||
```
Same as above comment. Maybe we should add externalLocation.startsWith("s3:")

kensipe replied:
+1 and resolved.

mohitsoni commented:
Looks good overall. Can we please add some tests ? I think specially for:
1. `PageBlobInputStream`
2. `PageBlobOutputStream`

Also, it would be great if you can please add some sanity tests for Azure (just like AWS), and maybe mark them as `azure`, so that they don't run by default. But, atleast we'll have an option to run them on demand OR in future when we'll have test automation on Azure.

kensipe committed:
added spock tests, added check for s3 on validity checks of backup restore requests and testing around it

kensipe committed:
removed some dead code.  added spock testing.  azure connection https now.  tested.

kensipe commented:
@mohitsoni added the validation checks and switched azure connection to https.  I added some testing for the validation checks.   The PageBlob unit tests are more complex based on final classes and final methods requiring something like powermocks.  It appears that powermocks have a dependency on an older version of mockito which when pulled into the project causes existing tests to fail.     It is unclear to me if powermock will work with the version of mockito we have in our project.    The sanity tests are a good idea but belong to another project.

kensipe commented:
After removing the conflicting test (which didn't test anything), still left with https://github.com/Azure/azure-sdk-for-java/issues/655 
I'm able to get around it, however then we run into an issue of mocking a final class that has no zero arg constructor.    Research indicates to use MSFT "fakes" to solve this but that isn't available for Java.

kensipe committed:
removing unnecessary test from the project

mohitsoni commented on cassandra-commons/build.gradle:
```diff
@@ -4,6 +4,7 @@ targetCompatibility = '1.8'
 apply plugin: 'com.google.protobuf'
 dependencies {
     compile "mesosphere:dcos-commons:0.1.0"
+    //     compile "mesosphere:dcos-commons:0.2.7-SNAPSHOT"  (needs to be upgraded to this, however it has breaking changes)
```
Can we remove this comment ? And, create a separate task to track this ?

kensipe replied:
resolved

mohitsoni commented on gradle/spock.gradle:
```diff
@@ -2,26 +2,26 @@
+  testCompile "org.powermock:powermock-module-junit4:$powermockVersion"
+  testCompile "org.powermock:powermock-module-junit4-rule:$powermockVersion"
+  testCompile "org.powermock:powermock-classloading-xstream:$powermockVersion"
+//  testCompile "org.powermock:powermock-api-mockito:$powermockVersion"
```
Can we please remove this too, if not required ?

kensipe replied:
resolved.

kensipe committed:
offending test was removed we can add powermock mockito back

kensipe committed:
removing dead code

mohitsoni commented:
This works, we verified the changes on a test cluster. 

Future work:
1. Add integration tests
2. Write unit tests for `PageBlobInputStream` and `PageBlobOutputStream`

Please raise JIRAs to track above.
